multi-arch/samples.md

```diff
@@ -0,0 +1,157 @@
+  - "@crawford"
+  - "@bparees"
+  - "@adambkaplan"
+approvers:
+  - "@smarterclayton"
+  - "@crawford"
+  - "@bparees"
+creation-date: 2019-09-20
+last-updated: 2019-09-20
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Sample Imagestream and Templates on non-x86 Platforms
+
+As of this writing, all samples are provided and owned by non-OpenShift 
+Development teams within Red Hat.  And none of those teams 
+
+ - provide host images on registry.redhat.io or quay.io that support PPC or Z 
+ - declare imagestreams in https://github.com/openshift/library that could reference such images
+
+OpenShift is now looking at providing support for the two non-x86 architectures from IBM in v4.x.
+
+Given that, until a resonable set of samples are provided for the IBM non-x86 architectures by either
+those other teams in Red Hat, or by IBM, the samples operator should ensure that it
+
+ - does not install any x86 imagestreams, or templates referencing those imagestreams, on non-x86 systems
+ - and by extension, always provides valid ClusterOperator statuses during install and upgrade, as this amounts to a no-op on those platforms
+
+And in case the implication is not clear, OpenShift development will not be developing or maintaining 
+any non-x86 images for these existing samples.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+As part of achieving what is described in the opening section, the samples operator needs
+to achieve the desired results within the context of how multi-arch is being achieved for 
+the entire product.
+
+Details around how mulit-arch will be achieved across the entire product are still pending,
+at least at the time of this writing for the author.
+
+In particular:
+ - will support across the entire product be staged along homogoneous / heterogeneous lines
+ - if "homogeneous only" becomes a valid option for the released product, is use of Golang's APIs for determining architecture sufficient
+ - if "heterogenous" is an option, the assumption is the list of valid architectures will be provided in a cluster-wide API object, and that samples operator installs for all architectures listed
+ - if "heterogenous" is an option, some notion outside the scope of the cluster operator API determines if a given sample runs on its associated architecture (for example, is a node selector needed on a deployment config or build leveraging a non-x86 imagestream to make sure the execution occurs on the right architecture?)
+
+## Motivation
+
+Multi-arch has been declared a key initiative for v4.x.
+
+### Goals
+
+x86 samples are not installed in a cluster with no x86 nodes.  Nor are any future non-x86 samples installed in a cluster with only x86 nodes.
```

Comment:
Do you have background motivation for this goal?

Reply:
the motivation is "they won't work"

Reply:
Does it matter if they won't work?  If we get heterogeneous clusters, I think we'll want to provide samples for all arches that exist in the cluster, so you can, for example, launch a Redis container without caring about the target node arch.  However we do that, I don't see a problem installing samples for arches that don't exist in the cluster too.

Reply:
When we are ready to implement/provide heterogeneous clusters we can revisit (I think i made a comment to that effect later in this proposal), until then, providing x86 samples on a non-x86 cluster that can't include x86 nodes is simply leading users into failure....and right now that failure reason is particularly unclear, as well (for which i've opened a BZ).  Basically if you try to run a ppc image on an x86 node, it will go into a crashloop and the logs/error message are not going to make it apparent why.

Note also that imagestreams do not support a mode of "give me a pod running this imagestreamtag and please pick the right architecture for the node i land on".  That's a manifestlist concept, and imagestreamtags represent specific image manifests, not a manifestlist.  So even if we did provide imagestreamtags that imported various architectures, the user would still have to pick the right imagestreamtag and align it with a nodeselector that steered their pod to the right node architecture.

Again, these are all things that need significant more discussion/design when we are prepared to define true heterogeneous support.

Reply:
I'd phrase this goal as:

> Install available samples for the current architecture. Do not install samples for other architectures while we only support homogenous clusters, to avoid currently-confusing error messages when an image for one architecture is run on a machine with a different architecture.

That motivates the current plan while making it clear that our concerns about samples from other arches are based on current technical limitations that we might resolve in the future.

Reply:
switched to that wording ... will push momentarily